invenio_accounts/tasks.py

```diff
@@ -73,3 +74,77 @@ def delete_ips():
         LoginInformation.current_login_at < expiration_date,
     ).update({LoginInformation.current_login_ip: None})
     db.session.commit()
+
+
+@shared_task
+def update_domain_status():
+    """Update domain statistics."""
+    # This subquery calculate the number of users per domain from the users
+    # table.
+    subquery = (
+        db.session.query(
+            User.domain,
+            func.count(User.id).label("num_users"),
+            func.count(User.active).filter(User.active == True).label("num_active"),
+            func.count(User.active).filter(User.active == False).label("num_inactive"),
+            func.count(User.confirmed_at).label("num_confirmed"),
+            func.count(User.verified_at).label("num_verified"),
+            func.count(User.blocked_at).label("num_blocked"),
+        )
+        .group_by(User.domain)
+        .subquery("n")
+    )
+
+    # Using above subquery, we find the domains that has changed.
+    stmt = (
+        db.session.query(
+            Domain.domain,
+            subquery.c.num_users,
+            subquery.c.num_active,
+            subquery.c.num_inactive,
+            subquery.c.num_confirmed,
+            subquery.c.num_verified,
+            subquery.c.num_blocked,
+        )
+        .join(subquery, Domain.domain == subquery.c.domain)
+        .filter(
+            or_(
+                Domain.num_users != subquery.c.num_users,
+                Domain.num_active != subquery.c.num_active,
+                Domain.num_inactive != subquery.c.num_inactive,
+                Domain.num_confirmed != subquery.c.num_confirmed,
+                Domain.num_verified != subquery.c.num_verified,
+                Domain.num_blocked != subquery.c.num_blocked,
+            )
+        )
+    )
+
+    # If statistics are updated regularly, the number of updates is relatively
+    # low and hence fit in memory. We read all data first, to avoid starting
+    # to modify the same table we're reading from.
+    domain_updates = []
+    for row in stmt.all():
+        domain_updates.append(row)
+
+    # Update the database
+    i = 0
+    now = datetime.utcnow()
+    for row in domain_updates:
+        domain, users, active, inactive, confirmed, verified, blocked = row
+        db.session.query(Domain).filter(Domain.domain == domain).update(
+            {
+                "num_users": users,
+                "num_active": active,
+                "num_inactive": inactive,
+                "num_confirmed": confirmed,
+                "num_verified": verified,
+                "num_blocked": blocked,
+                "updated": now,
+            }
+        )
+        i += 1
+        # Commit batches of 500 updates
+        if i % 500 == 0:
+            db.session.commit()
+    if i % 500 != 0:
+        db.session.commit()
```

Comment:
```suggestion
    domain_updates = stmt.all()

    # Commit batches of 500 updates
    batch_size = 500
    now = datetime.utcnow()

    # Process updates in batches
    for i in range(0, len(domain_updates), batch_size):
        with db.session.begin_nested():  # Use nested transactions for safety
            for domain, users, active, inactive, confirmed, verified, blocked in domain_updates[i:i + batch_size]:
                db.session.query(Domain).filter(Domain.domain == domain).update({
                    "num_users": users,
                    "num_active": active,
                    "num_inactive": inactive,
                    "num_confirmed": confirmed,
                    "num_verified": verified,
                    "num_blocked": blocked,
                    "updated": now,
                })
        db.session.commit() # Commit after each batch
```
no need for a final commit check outside the loop as it's handled within the loop.

Reply:
Should we try, expect any errors while unpacking the domain_updates, such as changes in the structure that could lead to a ValueError, or others?

Reply:
Updated. I would simply let the full task fail if there's an error and let the error logging take care of the rest. Main thing I can think of would be transaction conflicts or similar.